src/bson_util/async_decode.rs

```diff
@@ -0,0 +1,214 @@
+use bson::{oid, Array, Bson, DecoderError, DecoderResult, Document};
```

Comment:
It seems the decimal128 parts of it are missing. I'm assuming that's due to some weirdness with the feature flag. My concern with that is what if a decimal128 is included in a wire response? Will we not be able to read/write it?

As a side note, the decimal128 stuff appears to use `unsafe`, which is probably something we will want to change when we flesh it out.

Reply:
See my new comment below w.r.t. decimal128. I was talking more about for 1.0 than in this PR specifically.

re: getting this in the BSON library: Ideally we would want maintain a single imlementation of this stuff, so I guess it would have to be the async one. For sync, I guess `bson` spins up its own runtime and blocks on it similar to how trust-dns handles their sync work? also cc @mbroadst, since the BSON stuff seems to be something we hadn't considered in the design phase.

Reply:
I'm a bit on the fence about whether we want the sync BSON to wrap the async BSON like we do for the driver. The driver is already a pretty heavyweight dependency, so I don't think requiring sync users to pull in a async runtime is a huge deal, but BSON is much lighter weight, so I'd be a bit more hesitant to require the inclusion of an async runtime potentially just read and write a few bytes. As I mentioned in my earlier comment, right now I'm leaning more towards providing async functionality as either a feature flag or a separate crate. I'm not as worried about having to maintain two implementations for the BSON crate because the implementations would be a lot smaller (both due to the fact that the BSON crate is smaller than the driver and that most of the BSON crate is orthogonal to I/O).

Reply:
In Swift we decided that pulling in the runtime in the BSON library wouldn't be a huge deal, since a large majority of consumers of the BSON library will be consumers of the driver which already pulls it in. I'm unsure if that's also the case in the Rust world, but it is something to consider. Regardless, this seems like something that we should sketch out at least informally, even if the solution ends up being to just duplicate all the code under a feature flag.

Reply:
Given that we're not actually doing any blocking I/O here, could we use [`futures::io::AllowStdIo`](https://docs.rs/futures/0.3.4/futures/io/struct.AllowStdIo.html) to reduce the duplication?

Reply:
I don't think we can assume that there won't be able blocking I/O done; `decode_document` and `encode_document` are generic over the type of reader and writer they use, and they're public functions, so theoretically they could be used with something like a stream (which is in fact how we use them in the driver).

Reply:
> which is in fact how we use them in the driver

and that makes the driver really slow without buffering. i have strong susspicion that async machinery would make it even slower.

Reply:
As per offline discussion, we now buffer the entire document into a vec, and then subsequently use the existing sync decode implementation from `bson`.